CB-14187: (ios) Change the InAppBrowser to allow custom schemes

wvengen opened the pull request:
…Auth

<!--
Please make sure the checklist boxes are all checked before submitting the PR. The checklist
is intended as a quick reference, for complete details please see our Contributor Guidelines:

http://cordova.apache.org/contribute/contribute_guidelines.html

Thanks!
-->

### Platforms affected

- iOS

### What does this PR do?

Implements [CB-14187](https://issues.apache.org/jira/browse/CB-14187), which adds support for the `AllowedSchemes` preference and `customscheme` event for iOS.

### What testing has been done on this change?

Tested on an iPhone 5SE running iOS 11.4.

### Checklist
- [x] [Reported an issue](http://cordova.apache.org/contribute/issues.html) in the JIRA database
- [x] Commit message follows the format: "CB-3232: (android) Fix bug with resolving file paths", where CB-xxxx is the JIRA ID & "android" is the platform affected.
- [x] Added automated test coverage as appropriate for this change: as [CB-14013](https://issues.apache.org/jira/browse/CB-14013) (https://github.com/apache/cordova-plugin-inappbrowser/pull/263) did not include additional tests, I expect it to be unnecessary here as well.

brodybits commented:
The proposed changes definitely seem to be consistent with the corresponding functionality for Android from GH-263 ([CB-14013](https://issues.apache.org/jira/browse/CB-14013)). A couple more things that could help:
- a quick and easy app that I or someone else could use to test on both iOS and Android
- document the custom scheme support for both Android and iOS

wvengen commented:
Thanks for your review, @brodybits! I've added a test and some documentation.
I opted for a manual test only, since this needs a `config.xml` preference (even though I added that to the CONTRIBUTING document). It might be useful to add the preference to the regular Cordova documentation at some point in the future (not sure if that is desirable or not).

brodybits commented:
Thanks @wvengen. Unfortunately I had some real trouble getting a test demo to work on iOS. A quick demo project that shows this functionality working the same way on both Android and iOS (as closely as possible) would be really helpful.

wvengen commented:
Thanks for testing. Apparently I didn't test it enough, sorry. Still, for me it works on iOS, except that sometimes the alert _Result verified_ is shown twice. That is because, apparently, the loadstop event is being fired for customscheme as well. I don't see how that can be avoided, as in [`webViewDidFinishLoad`](https://github.com/apache/cordova-plugin-inappbrowser/blob/9cc7f69625363de9872f9dbc028b739f195a4f93/src/ios/CDVInAppBrowser.m#L463) (nor the [originating method](https://github.com/apache/cordova-plugin-inappbrowser/blob/9cc7f69625363de9872f9dbc028b739f195a4f93/src/ios/CDVInAppBrowser.m#L946)) there is no context about the URL that caused it.

My proposal:
1. Mention `loadstop` being called for `customscheme` in the documentation (an extra `iOS quirks` in [addEventListener](https://github.com/apache/cordova-plugin-inappbrowser#inappbrowseraddeventlistener)).
2. In the test, only navigate to the custom URL at the first time it happens.

brodybits commented:
I gotta say that I am still not able to get it to work on either Android or iOS in my test project at <https://github.com/brodybits/cordova-iab-custom-scheme-test-wip> (with 1 test link and 3 button tests). I suspect that I am doing something wrong; it would be best if you can issue a PR to correct me.

If I would add the proposed version of this plugin using the following command:
```
cordova plugin add https://github.com/q-m/cordova-plugin-inappbrowser#feature/allowedschemes-ios
```

then add iOS platform and then run from Xcode, I would get the following console log message on iOS when I click the test link or any test button:
```
2018-07-20 17:35:38.166670-0400 HelloCordova[12282:282185] Failed to load webpage with error: The URL can’t be shown
```

I get the following console log output if I click the test 3 button:

```
2018-07-20 18:05:31.146655-0400 HelloCordova[12987:301089] NSURLConnection finished with error - code -1002
2018-07-20 18:05:31.155721-0400 HelloCordova[12987:299914] Failed to load webpage with error: The URL can’t be shown
```

Test does not work for me on Android either.

If another Cordova member with more experience with this plugin can review and test I would be grateful.

wvengen commented:
It took me some experimentation to find the cause, but since the page you load is local, it is loaded in the Cordova WebView instead of InAppBrowser. Make it `cordova.InAppBrowser.open('iabpage.html', '_blank')` and the example should work.

Some small corrections in: https://github.com/wvengen/cordova-iab-custom-scheme-test-wip/commit/5917542781efd220745dca3172489ff38f0b95a7 (should work without) and https://github.com/wvengen/cordova-iab-custom-scheme-test-wip/commit/821652d41c22347eb50cb61ff97dba5a2b5abc0e (makes link work).

brodybits commented:
Thanks @wvengen I will take another look today.

brodybits commented on tests/tests.js:
```diff
@@ -693,4 +698,29 @@ exports.defineManualTests = function (contentEl, createActionButton) {
+            }
+        });
+        ref.addEventListener('customscheme', function (e) {
+            if (e && e.url === "custom://test") {
```
I just discovered today that `npm run eslint` has a few failures:

```
/Users/brodybits/Documents/cordova/cordova-plugin-inappbrowser/tests/tests.js
  714:32  error  Strings must use singlequote                    quotes
  715:1   error  Expected indentation of 16 spaces but found 14  indent
  715:15  error  'alert' is not defined                          no-undef
  717:1   error  Expected indentation of 16 spaces but found 14  indent
  717:15  error  'alert' is not defined                          no-undef
  722:13  error  'alert' is not defined                          no-undef

✖ 6 problems (6 errors, 0 warnings)
  3 errors, 0 warnings potentially fixable with the `--fix` option.
```

I can think of 2 possible solutions for `'alert' is not defined`:
- use `// eslint-disable-line no-undef`
- preferred: use `window.alert` (less ugly, more likely to work on Windows platform)

A nice bonus would be to fix `alert` --> `window.alert` throughout `tests.js`.

wvengen replied:
Thanks! I went for the bonus in 93077ef (hoping it will help motivate you to complete the review ;) ). Tested it still works (Android).

brodybits commented:
I pulled the changes from <https://github.com/wvengen/cordova-iab-custom-scheme-test-wip/commit/821652d41c22347eb50cb61ff97dba5a2b5abc0e> (master branch) into my master branch, doesn't seem to help on either Android or iOS. Gotta say I am getting ready to give up on this one. I just pushed one more change to <https://github.com/brodybits/cordova-iab-custom-scheme-test-wip> to use your `q-m:feature/allowedschemes-ios` branch for now, it would be great if you could fix the sample for the sake of the user community.

If you want another expert to take a look I suggest you send a request to: <mailto:dev@cordova.apache.org>

brodybits commented:
Thanks @wvengen for the update, fixing the eslint warnings, and fixing the alert calls. I will probably need 1-2 weeks to take another look. Apologies for the extra delays.

wvengen commented:
No problem, looking forward to getting this merged, somewhere in the coming weeks if all works out. :)

janpio commented:
And passing :) (context: https://github.com/apache/cordova-plugin-inappbrowser/issues/307#issuecomment-426404322)

brodybits commented:
@wvengen can you help resolve the conflicts?

wvengen commented:
Thanks! I'll be happy to rebase on master once #276 is sorted out.

janpio commented:
#276 is merged, so this can be taken care of now I think.

wvengen commented:
I will look at it, thanks.

wvengen commented:
Rebased. Tested on Android (7) and iOS (iPhone 8) emulator.

jonathanli2 commented:
it looks like this change is only for uiwebview. 
Currently wkwebiew does not handle customer url scheme, except for the app store url. Please refer to CDVWKInappbrowser.m's below method

- (BOOL)webView:(WKWebView*)theWebView decidePolicyForNavigationAction:(NSURLRequest*)request;

Should the change also handle custom url for wkwebview?

Thanks
Jonathan

timbru31 commented:
@wvengen could you possibly rebase your change, resolve the conflicts and see if the code has to be migrated/copied to the UIWebView and WKWebView? Thanks a lot!

janpio commented:
We can take care of (trivial) merge conflicts as this ourselves @timbru31 via the GitHub UI, no need to put the work on the original committer. Just did so.

timbru31 commented:
I believe it’s not just merge conflict, but adding the code to the WKWebView, too?

Sent from my iPhone

> On 24. Apr 2019, at 4:20 PM, Jan Piotrowski <notifications@github.com> wrote:
> 
> We can take care of (trivial) merge conflicts as this ourselves @timbru31 via the GitHub UI, no need to put the work on the original committer. Just did so.
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub, or mute the thread.

janpio commented:
Yes, that part of course has to be evaluated by @wvengen. 
(Maybe @dpa99c can help out there as well?)

wvengen commented:
Any help is appreciated here, as I'm not at all familiar with WKWebView, and I don't have that much time currently to look into that.

timbru31 commented on src/ios/CDVUIInAppBrowser.m:
```diff
@@ -439,6 +439,21 @@ - (BOOL)isValidCallbackId:(NSString *)callbackId
     return NO;
 }
 
+- (BOOL)isAllowedScheme:(NSString*)scheme
```
This need to be added to the WKWebView, too.
